File: leetcode/climbing_stairs/climbing_stairs.js
/**
 * @param {number} n
 * @return {number}
 */


// var climbStairs = function (n) {
//     const cache = new Map()

//     function climbStairsHelp(n) {
//         if (n < 2) {
//             return 1
//         }


//         if (!cache.has(n)) {
//             cache.set(n, climbStairsHelp(n - 1) + climbStairsHelp(n - 2))
//         }

//         return cache.get(n)
//     }

//     return climbStairsHelp(n)
// };


var climbStairs = function (n) {
    if (n < 2) {
        return 1
    }

    let prev = 1, curr = 2, next = 2;

    for (let index = 2; index < n; index++) {
        next = prev + curr
        prev = curr
        curr = next
    }

    return next
};
console.log(climbStairs(46));

File: leetcode/climbing_stairs/climbing_stairs.dart
class Solution {
  int climbStairs(int n) {
    if (n == 1) {
      return 1;
    }
    var n1 = 1;
    var n2 = 2;

    for (var i = 3; i < n + 1; i++) {
      var temp = n2;
      n2 = n1 + n2;
      n1 = temp;
    }

    return n2;
  }
}
